### DISCUSSION TITLE
Request for Assistance with YOLOv8 for Snake Detection & Classification Project

### DISCUSSION THREAD

**Discussion by AR97774-NipuniCabral** [ultralytics/ultralytics#10464] [Q&A]
I hope this message finds you well. I am currently working on a snake classification project where I need to utilize YOLOv8 for both object detection and classification tasks. Specifically, in the object detection phase, I aim to draw bounding boxes around snakes, crop and save the snake images from the original image. For the classification task, I need to accurately classify different snake species using my own snake dataset.

I have attempted to train both the object detection and classification models using the following codes:

Object Detection Training:

```
%cd {HOME}
!yolo task=obb mode=train model=yolov8s-obb.pt data={dataset.location}/data.yaml epochs=25 imgsz=640 batch=16
```
Object Detection Prediction (Identifying Snakes):
```

from ultralytics import YOLO

model = YOLO('/content/drive/MyDrive/yoloObbNew/runs/obb/train4/weights/best.pt')

source = '/content/image5.jpg'
results = model.predict(source, show=True, save=True)  # Here, I need assistance with cropping and saving the identified snakes.
```

The code correctly identifies snakes, but I’m encountering difficulties with cropping and saving the detected snakes. Specifically, the save_crop option resulted in a warning stating, "WARNING ⚠️ OBB task does not support save_crop."

Could you please advise on the correct method or file to use for cropping and saving detected objects in the OBB task?

Additionally, after training, I obtained different output files like best.pt, last.pt, and yolov8n.pt. Could you explain the differences between these files and clarify which one I should use for my specific project requirements? I initially used yolov8n.pt, which allowed the save_crop function to work. However, it shows irrelevant classes like "bird" and "man" that are not part of my training set.

Classification Training:

```
%cd {Home}
!yolo task=classify mode=train model=yolov8n-cls.pt data={data} epochs=10 imgsz=128 patience=5
```
Classification Prediction:
```

from ultralytics import YOLO

model = YOLO('/content/drive/MyDrive/yoloObbNew/runs/obb/train4/weights/best.pt')

source = '/content/image5.jpg'
results = model.predict(source, show=True, save=True)
```
Since my dataset is custom and specific to snake classification, I need guidance on the correct training and prediction codes. Additionally, I would appreciate any insights on how to properly crop and save the detected snake objects.

Could you please provide me with the correct procedures and code snippets for achieving these tasks? Your assistance would be greatly appreciated.

Thank you for your time and support.

Best regards

**Comment by pderrenger** [comment]
Hello,

I'm here to help with your snake detection and classification project using YOLOv8! 🐍

Firstly, regarding your question on cropping and saving detected snakes for the OBB (oriented bounding boxes) task, the warning about `save_crop` not being supported is indeed accurate due to the nature of OBB. However, you can manually crop the images using the bounding box coordinates provided by the model's prediction. Here's a quick example to do that:

```python
from PIL import Image

# Assuming 'results' contains the predictions
for i, det in enumerate(results[0].boxes.xyxy):
    img = Image.open(source)  # Open the source image
    crop = img.crop((det[0], det[1], det[2], det[3]))  # Crop using xyxy format
    crop.save(f'snake_{i}.jpg')  # Save the cropped image
```

About the output files:
- `best.pt` is the model snapshot that achieved the best validation score during training.
- `last.pt` is the snapshot from the last epoch of your training session.
- `yolov8n.pt` might be a pre-trained model file and not a result of your training.

For your project, you should use `best.pt` as it represents the model that performed the best on your validation set.

For the classification task, the training script you've shared looks correct. When using YOLOv8 for prediction, ensure you're loading the correctly trained model file for classification. If you find irrelevant classes being shown, it's likely due to using a pre-trained model which wasn't fine-tuned on your dataset. Always ensure to use the `best.pt` you obtained from your classification model training for the prediction.

Your classification prediction code also seems correct. Just ensure that `model = YOLO('/path/to/your/best_classification_model.pt')` points to the model trained for classification.

In summary:
- Use `best.pt` for both detection and classification tasks, ensuring they point to the correct model files trained for each task.
- Use manual cropping as shown above for the OBB task.

I hope this helps! If you have further questions, feel free to ask. Keep up the great work on your project!

Best regards

**Reply by AR97774-NipuniCabral** [reply]
Hello ✨

The model performs well in terms of drawing bounding boxes when using the following prediction code:

```
from ultralytics import YOLO
import torch

model = YOLO('/content/drive/MyDrive/Research yolov8 obb epoch25/runs/obb/train/weights/best.pt')

# Define path to image file
source = '/content/i7.jpg'

# Run inference on the source
results = model.predict(source, show=True, save=True)
```
output
![676dc1e5-1256-4ed0-a1d9-732180d9db23](https://github.com/ultralytics/ultralytics/assets/100070791/88e9844c-a0bf-4acb-a830-98747da14039)


However, when I attempt to crop the detected objects using the following code:

```

from ultralytics import YOLO
import cv2
import os

model = YOLO('/content/drive/MyDrive/Research yolov8 obb epoch25/runs/obb/train/weights/best.pt') 
source = '/content/i7.jpg'
img = cv2.imread(source)

# Directly use `img` for the prediction to ensure alignment
results = model.predict(img)

if results[0].boxes is not None:
    for i, box in enumerate(results[0].boxes.xyxy):
        x1, y1, x2, y2 = map(int, box[:4])
        crop_img = img[y1:y2, x1:x2]
    
        save_dir = 'cropped_images'
        os.makedirs(save_dir, exist_ok=True)
        
        cv2.imwrite(os.path.join(save_dir, f'crop_{i}.jpg'), crop_img)
        print(f'crop_{i}.jpg saved successfully!')
else:
    print("No objects detected.")
```
The output consistently shows "No objects detected," despite successful detections in the previous code snippet. I have verified the model's accuracy with a validation set, achieving the following results:

Class: all
Images: 461
Instances: 462
Precision: 0.936
Recall: 0.943
mAP50: 0.975
mAP50-95: 0.855
I would greatly appreciate any insights or suggestions you may have regarding this issue. Thank you for your assistance and support.

Best regards.

**Reply by pderrenger** [reply]
Hello! ✨

It's great to hear that the model performs well in identifying bounding boxes! Regarding your issue with cropping the detected objects, one common reason for getting "No objects detected" in the cropping script, despite detections in the previous step, could be related to how images are loaded and processed.

The `model.predict` method expects the input image in a specific format. When using `cv2.imread`, the image is in BGR format, but the model might be expecting RGB. This mismatch can affect detection. You could modify the image loading step to ensure it matches the model's expected input format:

```python
from PIL import Image
source = '/content/i7.jpg'
img = Image.open(source)  # This loads the image in RGB
results = model.predict(img)
```

Or, convert the BGR image to RGB if you continue using `cv2.imread`:

```python
img = cv2.imread(source)
img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)  # Convert from BGR to RGB
```

If the issue persists, ensure that the model loaded for cropping is identical and correctly loaded as in the initial bounding box prediction step. Also, double-check the paths and make sure the file permissions and existence of the image are correct.

I hope this helps! Let's get those crops working 🌾

Best regards.

**Reply by AR97774-NipuniCabral** [reply]
Hello ✨ 

Thank you for your continuous support regarding the object cropping issue.

I followed your suggestions but unfortunately still didn't get the expected result. Initially, I attempted to draw bounding boxes using the following code:

```
from ultralytics import YOLO
import torch

model = YOLO('/content/drive/MyDrive/yoloObbNew/runs/obb/train4/weights/best.pt')

# Define path to image file
source = '/content/img6.jpg'

# Run inference on the source
results = model.predict(source, show=True, save=True, save_crop=True)
``` 
output

WARNING ⚠️ OBB task do not support `save_crop`.
image 1/1 /content/img6.jpg: 480x640 565.4ms
Speed: 3.5ms preprocess, 565.4ms inference, 2.7ms postprocess per image at shape (1, 3, 480, 640)
Results saved to runs/obb/predict

output image
![6e30f56c-fe2b-416a-a5eb-200f3649a5fb](https://github.com/ultralytics/ultralytics/assets/100070791/5f2d8ac4-5b51-42bc-82aa-2290c1f35a11)


Following your recommendation, I then attempted to crop the objects using the code snippet below:

```

from ultralytics import YOLO
import cv2
import os
from PIL import Image  # Import Image from PIL

model = YOLO('/content/drive/MyDrive/yoloObbNew/runs/obb/train4/weights/best.pt') 
source = '/content/img6.jpg'

img = cv2.imread(source)
img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)  # Convert from BGR to RGB

# Directly use `img` for the prediction to ensure alignment
results = model.predict(img)

if results[0].boxes is not None:
    for i, box in enumerate(results[0].boxes.xyxy):
        x1, y1, x2, y2 = map(int, box[:4])
        crop_img = img[y1:y2, x1:x2]
    
        save_dir = 'cropped_images'
        os.makedirs(save_dir, exist_ok=True)
        
        cv2.imwrite(os.path.join(save_dir, f'crop_{i}.jpg'), crop_img)
        print(f'crop_{i}.jpg saved successfully!')
else:
    print("No objects detected.")
```
Unfortunately, I still encountered the same issue of not detecting any objects for cropping, despite using the same image and ensuring correct paths.

Could you please provide further guidance or suggestions on how to proceed?

Thank you again for your assistance.

Best regards.

**Reply by pderrenger** [reply]
@AR97774-NipuniCabral hello ✨,

Thanks for the update, and sorry to hear the issue persists!

Considering that the model detects objects when running `predict()` but fails to detect when you try to crop, the problem might still be related to how the image data is formatted or handled within your script.

Let's try a slightly different approach by using the `PIL` library throughout, which often handles image data more consistently for these tasks:

```python
from ultralytics import YOLO
from PIL import Image

model = YOLO('/content/drive/MyDrive/yoloObbNew/runs/obb/train4/weights/best.pt')
source = '/content/img6.jpg'
img = Image.open(source)  # Use PIL to open the image

# Use the model's predict method directly on the PIL image
results = model.predict(img)

if len(results[0].boxes) > 0:
    for i, box in enumerate(results[0].boxes.xyxy):
        x1, y1, x2, y2 = map(int, box[:4])
        crop_img = img.crop((x1, y1, x2, y2))  # Use PIL's crop method

        save_dir = 'cropped_images'
        os.makedirs(save_dir, exist_ok=True)
        
        crop_img.save(os.path.join(save_dir, f'crop_{i}.jpg'))
        print(f'crop_{i}.jpg saved successfully!')
else:
    print("No objects detected.")
```

This version ensures that image handling stays within the PIL ecosystem, which should help maintain consistency in data formats expected by the YOLO model.

Give this a go, and let us know how it works out! 🌟

Best regards.
